src/rust/runtime/memory/buffer/demibuffer.rs

```diff
@@ -0,0 +1,824 @@
+        }
+    }
+
+    // ----------------
+    // Public Functions
+    // ----------------
+
+    pub fn is_heap_allocated(&self) -> bool {
+        self.get_tag() == Tag::Heap
+    }
+
+    #[cfg(feature = "libdpdk")]
+    pub fn is_dpdk_allocated(&self) -> bool {
+        self.get_tag() == Tag::Dpdk
+    }
+
+    // Returns the length of the data stored in the DemiBuffer.
+    // Note that while we return a usize here (for convenience), the value is guaranteed to never exceed u16::MAX.
+    pub fn len(&self) -> usize {
+        self.as_metadata().data_len as usize
+    }
+
+    // Removes `nbytes` bytes from the beginning of the DemiBuffer chain.
+    // Note: If `nbytes` is greater than the length of the first segment in the chain, then this function will fail and
+    // return an error, rather than remove the remaining bytes from subsequent segments in the chain.  This is to match
+    // the behavior of DPDK's rte_pktmbuf_adj() routine.
+    pub fn adjust(&mut self, nbytes: u16) -> Result<(), Fail> {
+        // ToDo: Review having this "match", since MetaData and MBuf are laid out the same, these are equivalent cases.
+        match self.get_tag() {
+            Tag::Heap => {
+                let metadata: &mut MetaData = self.as_metadata();
+                if nbytes > metadata.data_len {
+                    return Err(Fail::new(libc::EINVAL, "tried to remove more bytes than are present"));
+                }
+                metadata.data_off += nbytes;
+                metadata.pkt_len -= nbytes as u32;
+                metadata.data_len -= nbytes;
+            },
+            #[cfg(feature = "libdpdk")]
+            Tag::Dpdk => {
+                // Safety: rte_pktmbuf_adj is a FFI, which is safe since we call it with an actual MBuf pointer.
+                if unsafe { rte_pktmbuf_adj(self.get_mbuf(), nbytes) } == ptr::null_mut() {
+                    return Err(Fail::new(libc::EINVAL, "tried to remove more bytes than are present"));
+                }
+            },
+        }
+
+        Ok(())
+    }
+
+    // Removes `nbytes` bytes from the end of the DemiBuffer chain.
+    // Note: If `nbytes` is greater than the length of the last segment in the chain, then this function will fail and
+    // return an error, rather than remove the remaining bytes from subsequent segments in the chain.  This is to match
+    // the behavior of DPDK's rte_pktmbuf_trim() routine.
+    pub fn trim(&mut self, nbytes: u16) -> Result<(), Fail> {
+        // ToDo: Review having this "match", since MetaData and MBuf are laid out the same, these are equivalent cases.
+        match self.get_tag() {
+            Tag::Heap => {
+                let md_first: &mut MetaData = self.as_metadata();
+                let md_last: &mut MetaData = md_first.get_last_segment();
+
+                if nbytes > md_last.data_len {
+                    return Err(Fail::new(libc::EINVAL, "tried to remove more bytes than are present"));
+                }
```

Comment:
This check (and similar ones in `adjust` and `split`) either needs to be done separately for each case, or we need to unequivocally state (i.e., rely upon) that our `MetaData` struct and DPDK's `rte_mbuf` struct are laid out the same way (i.e., that the `data_len` value is at the same offset and has the same size for both).

I don't mind going "all in" on `MetaData`/`rte_mbuf` but was hesitant to do that originally because of the lack of compile-time assertions to check that these stay the same (roughly).  Now that I know how to do that reasonably, I think we can do this.  If we do though, we might as well remove all the special case code for DPDK in `adjust`/`trim` and just use our Rust implementations as I wrote them to be exactly equivalent to `rte_pktmbuf_adj` and `rte_pktmbuf_trim` and making the FFI calls out to DPDK code is undoubtedly less performant.

If we leave them separate, then I think it's okay to rely on DPDK's check because it literally does the same thing.

Reply:
> If we do though, we might as well remove all the special case code for DPDK in adjust/trim and just use our Rust implementations as I wrote them to be exactly equivalent to rte_pktmbuf_adj and rte_pktmbuf_trim and making the FFI calls out to DPDK code is undoubtedly less performant

Are you suggesting that we wouldn't call `rte_pktmbuf_adj` and `rte_pktmbuf_trim` at all? I wouldn't recommending doing so, because we would be assuming that future versions of DPDK wouldn't break this assumption.

Reply:
Yes, that was what I was suggesting.  In general, I don't think we should attempt to future-proof our code to work with any possible changes to DPDK.  Something like this would be a breaking change for DPDK, so I doubt they would do that.  And if they did, it would likely be coupled with a change to `rte_mbuf`, which would break us similarly.

I propose we leave this as-is for now and revisit it if the DPDK FFI calls turn out to be performance issues.

Reply:
I agree to leave as it is. However, I want to note that I am not in favor of dropping DPDK FFI calls for sake or performance, nor assuming that DPDK interface/internals won't work. DPDK is not a standard and it evolves rapidly. There are many features that 3 years ago simply weren't there.